groups/ntc/ntco/ntco_ioring.cpp

```diff
@@ -0,0 +1,6359 @@
+                                      static_cast<unsigned int>(completions),
+                                      flags,
+                                      reinterpret_cast<sigset_t*>(0),
+                                      _NSIG / 8));
+}
+
+int IoRingUtil::enter(int                       ring,
+                      bsl::size_t               submissions,
+                      bsl::size_t               completions,
+                      const bsls::TimeInterval& deadline)
+{
+    const long          k_SYSTEM_CALL_ENTER                = 426;
+    const bsl::uint32_t k_SYSTEM_CALL_ENTER_FLAG_GETEVENTS = 1U << 0;
+    const bsl::uint32_t k_SYSTEM_CALL_ENTER_FLAG_EXT_ARG   = 1U << 3;
+
+    bsl::uint32_t flags = k_SYSTEM_CALL_ENTER_FLAG_EXT_ARG;
+    if (NTCCFG_LIKELY(completions > 0)) {
+        flags |= k_SYSTEM_CALL_ENTER_FLAG_GETEVENTS;
+    }
+
+    const bsls::TimeInterval now = bdlt::CurrentTime::now();
+
+    bsls::TimeInterval duration;
+    if (NTCCFG_LIKELY(deadline > now)) {
+        duration = deadline - now;
+    }
+
+    struct __kernel_timespec timespec;
+    timespec.tv_sec  = duration.seconds();
+    timespec.tv_nsec = duration.nanoseconds();
+
+    struct IoRingOptions {
+        bsl::uint64_t signalMask;
+        bsl::uint32_t signalMaskSize;
+        bsl::uint32_t pad;
+        bsl::uint64_t timespec;
+    } options;
+
+    options.signalMask     = 0;
+    options.signalMaskSize = 0;
+    options.pad            = 0;
+    options.timespec       = reinterpret_cast<bsl::uint64_t>(&timespec);
+
+    return static_cast<int>(::syscall(k_SYSTEM_CALL_ENTER,
+                                      ring,
+                                      static_cast<unsigned int>(submissions),
+                                      static_cast<unsigned int>(completions),
+                                      flags,
+                                      &options,
+                                      sizeof options));
+}
+
+int IoRingUtil::control(int         ring,
+                        bsl::size_t operation,
+                        void*       operand,
+                        bsl::size_t count)
+{
+    const long k_SYSTEM_CALL_REGISTER = 427;
+
+    return static_cast<int>(::syscall(k_SYSTEM_CALL_REGISTER,
+                                      ring,
+                                      static_cast<unsigned int>(operation),
+                                      operand,
+                                      static_cast<unsigned int>(count)));
+}
+
+int IoRingUtil::probe(int ring, ntco::IoRingProbe* probe)
+{
+    const long          k_SYSTEM_CALL_REGISTER       = 427;
+    const bsl::uint32_t k_SYSTEM_CALL_REGISTER_PROBE = 8;
+
+    return static_cast<int>(::syscall(k_SYSTEM_CALL_REGISTER,
+                                      ring,
+                                      k_SYSTEM_CALL_REGISTER_PROBE,
+                                      probe,
+                                      256));
```

Comment:
better to replace with `IoRingProbe::k_ENTRY_COUNT` or pas as a parameter to the function

Reply:
OK